src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoroelasticKernel.hpp

```diff
@@ -0,0 +1,447 @@
+            xLocal(),
+            u_local(),
+            uhat_local(),
+            localFlowResidual{ 0.0 },
+      localDispFlowJacobian{ {0.0} },
+      localFlowDispJacobian{ {0.0} },
+      localFlowFlowJacobian{ {0.0} },
+      localFlowDofIndex{ 0 }
+    {}
+
+#if !defined(CALC_FEM_SHAPE_IN_KERNEL)
+    /// Dummy
+    int xLocal;
+#else
+    /// C-array stack storage for element local the nodal positions.
+    real64 xLocal[ numNodesPerElem ][ 3 ];
+#endif
+
+    /// Stack storage for the element local nodal displacement
+    real64 u_local[numNodesPerElem][numDofPerTrialSupportPoint];
+
+    /// Stack storage for the element local nodal incremental displacement
+    real64 uhat_local[numNodesPerElem][numDofPerTrialSupportPoint];
+
+    real64 localFlowResidual[1];
+    real64 localDispFlowJacobian[numDispDofPerElem][1];
+    real64 localFlowDispJacobian[1][numDispDofPerElem];
+    real64 localFlowFlowJacobian[1][1];
+
+    /// C-array storage for the element local row degrees of freedom.
+    globalIndex localFlowDofIndex[1];
+
+  };
+  //*****************************************************************************
+
+  /**
+   * @brief Copy global values from primary field to a local stack array.
+   * @copydoc ::geosx::finiteElement::ImplicitKernelBase::setup
+   *
+   * For the SinglePhase implementation, global values from the displacement,
+   * incremental displacement, and degree of freedom numbers are placed into
+   * element local stack storage.
+   */
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void setup( localIndex const k,
+              StackVariables & stack ) const
+  {
+    for( localIndex a=0; a<numNodesPerElem; ++a )
+    {
+      localIndex const localNodeIndex = m_elemsToNodes( k, a );
+
+      for( int i=0; i<3; ++i )
+      {
+#if defined(CALC_FEM_SHAPE_IN_KERNEL)
+        stack.xLocal[ a ][ i ] = m_X[ localNodeIndex ][ i ];
+#endif
+        stack.u_local[ a ][i] = m_disp[ localNodeIndex ][i];
+        stack.uhat_local[ a ][i] = m_uhat[ localNodeIndex ][i];
+        stack.localRowDofIndex[a*3+i] = m_dofNumber[localNodeIndex]+i;
+        stack.localColDofIndex[a*3+i] = m_dofNumber[localNodeIndex]+i;
+      }
+    }
+
+    for( int flowDofIndex=0; flowDofIndex<1; ++flowDofIndex )
+    {
+      stack.localFlowDofIndex[flowDofIndex] = m_flowDofNumber[k] + flowDofIndex;
+    }
+
+  }
+
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void quadraturePointKernel( localIndex const k,
+                              localIndex const q,
+                              StackVariables & stack ) const
+  {
+    // Get displacement: (i) basis functions (N), (ii) basis function
+    // derivatives (dNdX), and (iii) determinant of the Jacobian transformation
+    // matrix times the quadrature weight (detJxW)
+    real64 N[numNodesPerElem];
+    real64 dNdX[numNodesPerElem][3];
+    FE_TYPE::calcN( q, N );
+    real64 const detJxW = m_finiteElementSpace.template getGradN< FE_TYPE >( k, q, stack.xLocal, dNdX );
```

Comment:
This function is called getGradN but it return also detJxW, ce we optimize this?

Reply:
What do you mean by optimizing?

Reply:
I mean: 
- Option 1: the function name should reflect the fact that it is a multi task function that modifies gradN and return detJxW
- Option 2: this multi task function should be splitted to single purpose functions

Persionally I prefer single purpose functions

Reply:
Computing `dNdX` requires the Jacobian matrix of the isoparametric mapping. It is convenient to return its determinant, which is needed in the numerical integration. Single purpose functions are nice but we would end up recomputing over and over again the same quantities. We could change the function to something like:
```
getGradN< FE_TYPE >( k, q, stack.xlocal, dNdX, detJxW );
```
Personally I think the return value (`detJxW`) makes sense. But I'm open to suggestions.

Reply:
I think this would hide the fact that `detJxW ` gets modified inside the function so I kind of prefere the way it currently is.